packages/SwingSet/docs/vat-environment.md

```diff
@@ -0,0 +1,113 @@
+
+* `import` and `export` syntax are allowed in ESM-style modules (which is preferred over CommonJS). These are not globals per se, but rather top-level syntax which defines the module graph.
+* `require`, `module`, `module.exports`, and `exports` are allowed in CommonJS-style modules, and should work as expected, however new code should be written as ESM modules. They are either consumed by the the bundling process, provided (in some form) by the execution environment, or otherwise rewritten to work sensibly
+* `__dirname` and `__filename` are not provided
+* The dynamic import expression (`await import('name')`) is currently prohibited in vat code, but a future SES implementation may allow it.
+
+Node.js has a [large collection](https://nodejs.org/dist/latest-v14.x/docs/api/) of "built-in modules", like `http` and `crypto`. Some are clearly platform-specific (like `v8`), while others are not so obvious (`stream`). All of these are accessed by importing a module (`const v8 = require('v8')` in CommonJS modules, or `import v8 from 'v8'` in ESM modules). These modules are built out of native code (C++), not plain JS.
+
+None of these built-in modules are available to Vat code. `require` or `import` can be used on modules that contain pure JS, but not on modules that include native code. To allow a vat to exercise authority which comes from a built-in module, you will have to write a "device", give that device an endowment with the built-in module's functions, then have the vat send messages to the device.
+
+Browser environments also have a huge list of [other features](https://developer.mozilla.org/en-US/docs/Web/API) which are presented as names in the global scope (some of which were also added to Node.js). None of these are available in a SES environment. The most surprising removals include `atob`, `TextEncoder`, and `URL`.
+
+`debugger` is a first-class JavaScript statement, and behaves as expected in vat code.
+
+## Other Changes
+
+The [shim](https://github.com/Agoric/SES-shim/) which provides our SES environment is not as fully-featured as a native implemenation could be. As a result, some forms of code cannot be used yet. These restrictions should be lifted once the JS engine you use can provide SES natively:
+
+### HTML Comments
+
+JavaScript parsers may or may not recognize HTML comments within source code, potentially leading to different behavior on different engines. For safety, the SES shim rejects any source code which looks like it might contain a comment open (`<!--`) or close (`-->`) sequence. The filter uses a regular expression, not a full parser.
```

Comment:
"may or may not" -> "may not"

"looks like" or does contain? If the latter, change to "which contains a comment open..."

Reply:
I wanted to emphasize the uncertainty in both cases. The JS spec doesn't require any specific behavior, so it's a complete toss-up what any given engine does with those comments. What about "Some JavaScript parsers will recognize HTML comments, and ignore any code within them. Other parsers will not, potentially.." ?

The shim's filter is not very precise. It will reject code that does not *actually* contain a comment sequence (i.e. the filter regexp will reject `const commentOpen = '<!--';`, even though a JS parser might be smart enough to realize that this is a string-literal). I was aiming for "anything that bears a passing resemblance to a comment sequence", but that came out too verbose :). Maybe `any source code which even *looks* like it might..` ? Or "which might possibly contain .." ?

Reply:
I'm concerned about the vagueness of the comment open matching. If "<!---" is a comment open, what's it likely to get wrong? "anything like" could mean, for example, anything with "---". Or you need watch out for "<"s. Or five non-alphanumerics in a row. Or would "Any source code containing the string `<!---`, even if it doesn't actually start a comment, ..." work?

Reply:
the latter sounds right, I'll do that

Reply:
JavaScript parsers may or may not recognize HTML comments within source code, potentially leading to different behavior on different engines. For safety, the SES shim rejects any source code which contains a comment open (`<!--`) or close (`-->`) sequence. However the filter uses a regular expression, not a full parser, so it will unnecessarily reject any source code containing the string `<!--` (or `-->`), even if it does not actually mark a comment.